dali/pipeline/operators/crop/crop.cc

```diff
@@ -29,57 +40,57 @@ DALI_SCHEMA(Crop)
     .AddOptionalArg("crop_pos_y",
                     R"code(Vertical position of the crop in image coordinates (0.0 - 1.0))code",
                     0.5f, true)
-    .AddOptionalArg("image_type",
-                    R"code(The color space of input and output image)code", DALI_RGB, false)
     .AddArg("crop",
             R"code(Size of the cropped image. If only a single value `c` is provided,
  the resulting crop will be square with size `(c,c)`)code", DALI_INT_VEC)
+    .AddParent("CastPermute")
     .EnforceInputLayout(DALI_NHWC);
 
 
 template<>
-Crop<CPUBackend>::Crop(const OpSpec &spec) : Operator<CPUBackend>(spec), CropAttr(spec) {
+Crop<CPUBackend>::Crop(const OpSpec &spec, bool defaultCastPermut) :
+                  Operator<CPUBackend>(spec), CropAttr(spec, defaultCastPermut) {
   Init(num_threads_);
 }
 
-template<typename Out>
-void CropKernel(
-  const int C,
-  const int H,
-  const int W,
-  const unsigned char *input_ptr,
-  const int in_stride,
-  DALITensorLayout layout,
-  Out *output_ptr) {
-  if (layout == DALI_NCHW) {
-    for (int c = 0; c < C; ++c) {
-      for (int h = 0; h < H; ++h) {
-        for (int w = 0; w < W; ++w) {
-          // From HWC
-          const int in_idx = h * in_stride + w * C + c;
-          // To CHW
-          const int out_idx = (c * H + h) * W + w;
-          output_ptr[out_idx] = static_cast<Out>(input_ptr[in_idx]);
-        }
-      }
-    }
-  } else {  // Layout == DALI_NHWC
-    for (int c = 0; c < C; ++c) {
-      for (int h = 0; h < H; ++h) {
-        for (int w = 0; w < W; ++w) {
-          // From HWC
-          const int in_idx = h * in_stride + w * C + c;
-          // To HWC
-          const int out_idx = (h * W + w) * C + c;
-          output_ptr[out_idx] = static_cast<Out>(input_ptr[in_idx]);
-        }
-      }
-    }
-  }
+#define CROP_KERNEL(ext, Out, conv1, conv2)                 \
```

Comment:
Why it needs to be a macro instead of template?

Reply:
I was trying to use code which would be used in both cases: for intrinsics (`_cvtss_sh(x, 0)`)AND for previously used `static_cast<float16>(x)` and all other types. Cliff showed me better alternative  to reach this goal - the functors used the templates. My new commit will have these changes.